src/node_quic_buffer.h

```diff
@@ -0,0 +1,255 @@
+#ifndef SRC_NODE_QUIC_BUFFER_H
+#define SRC_NODE_QUIC_BUFFER_H
+
+#if defined(NODE_WANT_INTERNALS) && NODE_WANT_INTERNALS
+
+#include "node.h"
+#include "node_internals.h"
+#include "uv.h"
+#include "v8.h"
+
+#include <algorithm>
+#include <functional>
+#include <vector>
+
+namespace node {
+namespace quic {
+
+// QuicBuffer an internal linked list of uv_buf_t instances
+// representing data that is to be sent. All data in the
+// Buffer has to be retained until it is Consumed or Canceled.
+// For QUIC, the data is not consumed until an explicit ack
+// is received.
+
+typedef std::function<void(int status, void* user_data)> done_cb;
+
+// Default non-op done handler.
+void default_quic_buffer_chunk_done(int status, void* user_data) {}
+
+struct quic_buffer_chunk : public MemoryRetainer {
+  uv_buf_t buf;
+  done_cb done = default_quic_buffer_chunk_done;
+  size_t offset = 0;
+  void* user_data = nullptr;
+  bool done_called = false;
+  v8::Global<v8::Object> keep_alive;
+  std::unique_ptr<quic_buffer_chunk> next;
+
+  inline explicit quic_buffer_chunk(
+    uv_buf_t buf_) :
+    buf(buf_) {}
+
+  inline quic_buffer_chunk(
+    uv_buf_t buf_,
+    done_cb done_,
+    void* user_data_,
+    v8::Local<v8::Object> keep_alive_) :
+    buf(buf_),
+    done(done_),
+    user_data(user_data_) {
+    if (!keep_alive.IsEmpty())
+      keep_alive.Reset(keep_alive_->GetIsolate(), keep_alive_);
+  }
+
+  inline ~quic_buffer_chunk() {
+    CHECK(done_called);
+  }
+
+  void Done(int status) {
+    done_called = true;
+    done(status, user_data);
+  }
+
+  void MemoryInfo(MemoryTracker* tracker) const override {
+    tracker->TrackFieldWithSize("buf", buf.len);
+  }
+
+  SET_MEMORY_INFO_NAME(quic_buffer_chunk)
+  SET_SELF_SIZE(quic_buffer_chunk)
+};
+
+class QuicBuffer : public MemoryRetainer {
+ public:
+  inline QuicBuffer() :
+    head_(nullptr),
+    tail_(nullptr),
+    size_(0),
+    length_(0) {}
+
+  inline ~QuicBuffer() {
+    Cancel();  // Cancel the remaining data
+    // Verify that all data has been purged, either through
+    // Consume or Cancel
+    CHECK_EQ(length_, 0);
+  }
+
+  // Transfers the state of this QuicBuffer to another
+  // instance, including ownership of the linked list
+  // of uv_buf_t instances. The destination must be
+  // an empty QuicBuffer.
```

Comment:
Does it make sense to implement this as a move constructor? The function source looks exactly like one :)

Reply:
yep, was thinking the same.